[2.10] - Fields containing {{type:Back}} no longer render

Reported on reddit

* Note Type: Basic

Front:
```
AA{{type:Back}}
```

* In 2.9, this renders as a "Type Answer" card.

* In 2.10, this shows `AA{{type:Back}}`

Regression test, will update with a bisect ASAP

<details><summary>TemplateTest.java</summary>

```
/*
 Copyright (c) 2020 David Allison <davidallisongithub@gmail.com>

 This program is free software; you can redistribute it and/or modify it under
 the terms of the GNU General Public License as published by the Free Software
 Foundation; either version 3 of the License, or (at your option) any later
 version.

 This program is distributed in the hope that it will be useful, but WITHOUT ANY
 WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 PARTICULAR PURPOSE. See the GNU General Public License for more details.

 You should have received a copy of the GNU General Public License along with
 this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.ichi2.libanki;

import com.ichi2.libanki.template.Template;

import org.junit.Test;
import org.junit.runner.RunWith;

import java.util.HashMap;

import androidx.test.ext.junit.runners.AndroidJUnit4;

import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.is;

@RunWith(AndroidJUnit4.class)
public class TemplateTest {

    @Test
    public void testDataInField() {
        HashMap<String, String> context = new HashMap<>();
        context.put("Front", "AA{{type:Back}}");
        Template t = new Template("{{Front}}", context);

        String rendered = t.render();

        assertThat(rendered, is("AA[[type:Back]]"));
    }
}

```

</details>

<details><summary>Stack Trace</summary>

java.lang.Exception: Main looper has queued unexecuted runnables. This might be the cause of the test failure. You might need a shadowOf(getMainLooper()).idle() call.

	at org.robolectric.android.internal.AndroidTestEnvironment.checkStateAfterTestFailure(AndroidTestEnvironment.java:470)
	at org.robolectric.RobolectricTestRunner$HelperTestRunner$1.evaluate(RobolectricTestRunner.java:548)
	at org.robolectric.internal.SandboxTestRunner$2.lambda$evaluate$0(SandboxTestRunner.java:252)
	at org.robolectric.internal.bytecode.Sandbox.lambda$runOnMainThread$0(Sandbox.java:89)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.AssertionError: 
Expected: is "AA[[type:Back]]"
     but: was "AA{{type:Back}}"
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:20)
	at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:8)
	at com.ichi2.libanki.TemplateTest.testDataInField(TemplateTest.java:42)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.robolectric.RobolectricTestRunner$HelperTestRunner$1.evaluate(RobolectricTestRunner.java:546)
	... 6 more
```
</details>

--- a/AnkiDroid/src/main/java/com/ichi2/libanki/template/Template.java
+++ b/AnkiDroid/src/main/java/com/ichi2/libanki/template/Template.java
@@ -31,6 +31,7 @@ import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
 import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
 import timber.log.Timber;
 
 /**
@@ -72,7 +73,7 @@ public class Template {
     private Map<String, String> mContext;
 
 
-    private static String get_or_attr(Map<String, String> obj, String name) {
+    private static @Nullable String get_or_attr(Map<String, String> obj, String name) {
         if (obj.containsKey(name)) {
             return obj.get(name);
         } else {
@@ -81,7 +82,7 @@ public class Template {
     }
 
 
-    public Template(String template, Map<String, String> context) {
+    public Template(@NonNull String template, @Nullable Map<String, String> context) {
         mTemplate = template;
         mContext = context == null ? new HashMap<>() : context;
     }
@@ -90,7 +91,7 @@ public class Template {
     /**
      * Turns a Mustache template into something wonderful.
      */
-    public String render() {
+    public @NonNull String render() {
         String template = render_sections(mTemplate, mContext);
         return render_tags(template, mContext);
     }
@@ -98,7 +99,7 @@ public class Template {
     /**
      * Expands sections.
      */
-    private String render_sections(@NonNull String template, @NonNull Map<String, String> context) {
+    private @NonNull String render_sections(@NonNull String template, @NonNull Map<String, String> context) {
         /* Apply render_some_section to the templates, until
            render_some_section states that it does not find sections
            anymore. Return the last template found. */
@@ -127,7 +128,7 @@ public class Template {
      If some change is done, the function should be called again to
      remove those new pairs of conditionals.
      */
-    private String render_some_sections(@NonNull String template, @NonNull Map<String, String> context) {
+    private @Nullable String render_some_sections(@NonNull String template, @NonNull Map<String, String> context) {
         StringBuffer sb = new StringBuffer();
         Matcher match = sSection_re.matcher(template);
         boolean found = false;
@@ -157,7 +158,7 @@ public class Template {
     /**
      * Expands all tags, iteratively until all tags (even tags that are replaced by tags) are resolved.
      */
-    private String render_tags(@NonNull String template, @NonNull Map<String, String> context) {
+    private @NonNull String render_tags(@NonNull String template, @NonNull Map<String, String> context) {
         /* Apply render_some_tags to the tags, until
            render_some_tags states that it does not find tags to replace anymore
            anymore. Return the last template state */
@@ -172,7 +173,7 @@ public class Template {
     /**
      * Replaces all the tags in a template in a single pass for the values in the given context map.
      */
-    private String render_some_tags(String template, Map<String, String> context) {
+    private @Nullable String render_some_tags(@NonNull String template, @NonNull Map<String, String> context) {
         String ALT_HANDLEBAR_DIRECTIVE = "{{=<% %>=}}";
         if (template.contains(ALT_HANDLEBAR_DIRECTIVE)) {
             template = template.replace(ALT_HANDLEBAR_DIRECTIVE, "").replace("<%", "{{").replace("%>", "}}");
@@ -206,7 +207,7 @@ public class Template {
     /**
      * {{{ functions just like {{ in anki
      */
-    private String render_tag(String tag_name, Map<String, String> context) {
+    private @NonNull String render_tag(@NonNull String tag_name, @NonNull Map<String, String> context) {
         return render_unescaped(tag_name, context);
     }
 
@@ -218,7 +219,7 @@ public class Template {
         return "";
     }
 
-    private String render_unescaped(String tag_name, Map<String, String> context) {
+    private @NonNull String render_unescaped(@NonNull String tag_name, @NonNull Map<String, String> context) {
         String txt = get_or_attr(context, tag_name);
         if (txt != null) {
             // some field names could have colons in them
@@ -279,7 +280,7 @@ public class Template {
                 mod = split[0];
                 extra = split[1];
                 if (!TextUtils.isEmpty(txt) && !TextUtils.isEmpty(extra)) {
-                    txt = clozeText(txt, extra, mod.charAt(1));
+                    txt = clozeText(txt != null ? txt : "", extra, mod.charAt(1));
                 } else {
                     txt = "";
                 }
@@ -300,10 +301,10 @@ public class Template {
                 }
             }
         }
-        return txt;
+        return txt != null ? txt : "";
     }
 
-    private static String clozeText(String txt, String ord, char type) {
+    private static @NonNull String clozeText(@NonNull String txt, @NonNull String ord, char type) {
         if (!Pattern.compile(String.format(Locale.US, clozeReg, ord)).matcher(txt).find()) {
             return "";
         }
@@ -336,7 +337,7 @@ public class Template {
         return txt.replaceAll(String.format(Locale.US, clozeReg, "\\d+"), "$2");
     }
 
-    public static boolean textContainsMathjax(String txt) {
+    public static boolean textContainsMathjax(@NonNull String txt) {
         // Do you have the first opening and then the first closing,
         // or the second opening and the second closing...?
 
@@ -375,7 +376,7 @@ public class Template {
      * The clozeText method interprets the upper-case C as "don't wrap this
      * Cloze in a <span>".
      */
-    public static String removeFormattingFromMathjax(String txt, String ord) {
+    public static @NonNull String removeFormattingFromMathjax(@NonNull String txt, @NonNull String ord) {
         String creg = clozeReg.replace("(?si)", "");
         // Scan the string left to right.
         // After a MathJax opening - \( or \[ - flip in_mathjax to True.

--- a/AnkiDroid/src/main/java/com/ichi2/libanki/template/Template.java
+++ b/AnkiDroid/src/main/java/com/ichi2/libanki/template/Template.java
@@ -17,14 +17,12 @@
 package com.ichi2.libanki.template;
 
 import android.text.TextUtils;
-import android.util.Log;
 
 import com.ichi2.libanki.Utils;
 import com.ichi2.libanki.hooks.Hooks;
 
 import java.util.Arrays;
 import java.util.Collections;
-import java.util.Comparator;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Locale;
@@ -33,6 +31,7 @@ import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
 import androidx.annotation.NonNull;
+import timber.log.Timber;
 
 /**
  * This class renders the card content by parsing the card template and replacing all marked sections
@@ -50,7 +49,6 @@ import androidx.annotation.NonNull;
 public class Template {
     public static final String clozeReg = "(?si)\\{\\{(c)%s::(.*?)(::(.*?))?\\}\\}";
     private static final Pattern fHookFieldMod = Pattern.compile("^(.*?)(?:\\((.*)\\))?$");
-    private static final String TAG = Template.class.getName();
 
     // Opening tag delimiter
     private static final String sOtag = Pattern.quote("{{");
@@ -59,16 +57,16 @@ public class Template {
     private static final String sCtag = Pattern.quote("}}");
 
     // The regular expression used to find a #section
-    private static final Pattern sSection_re = Pattern.compile(sOtag + "[\\#|^]([^\\}]*)" + sCtag + "(.+?)" + sOtag + "/\\1" + sCtag, Pattern.MULTILINE | Pattern.DOTALL);
+    private static final Pattern sSection_re = Pattern.compile(sOtag + "[#|^]([^}]*)" + sCtag + "(.+?)" + sOtag + "/\\1" + sCtag, Pattern.MULTILINE | Pattern.DOTALL);
 
     // The regular expression used to find a tag.
-    private static final Pattern sTag_re = Pattern.compile(sOtag + "(#|=|&|!|>|\\{)?(.+?)\\1?" + sCtag + "+");
+    private static final Pattern sTag_re = Pattern.compile(sOtag + "([#=&!>{])?(.+?)\\1?" + sCtag + "+");
 
     // MathJax opening delimiters
-    private static String sMathJaxOpenings[] = {"\\(", "\\["};
+    private static String[] sMathJaxOpenings = {"\\(", "\\["};
 
     // MathJax closing delimiters
-    private static String sMathJaxClosings[] = {"\\)", "\\]"};
+    private static String[] sMathJaxClosings = {"\\)", "\\]"};
 
     private String mTemplate;
     private Map<String, String> mContext;
@@ -76,21 +74,17 @@ public class Template {
     private static String ALT_HANDLEBAR_DIRECTIVE = "{{=<% %>=}}";
 
     private static String get_or_attr(Map<String, String> obj, String name) {
-        return get_or_attr(obj, name, null);
-    }
-
-    private static String get_or_attr(Map<String, String> obj, String name, String _default) {
         if (obj.containsKey(name)) {
             return obj.get(name);
         } else {
-            return _default;
+            return null;
         }
     }
 
 
     public Template(String template, Map<String, String> context) {
         mTemplate = template;
-        mContext = context == null ? new HashMap<String, String>() : context;
+        mContext = context == null ? new HashMap<>() : context;
     }
 
 
@@ -143,7 +137,7 @@ public class Template {
             String section = match.group(0);
             String section_name = match.group(1).trim();
             String inner = match.group(2);
-            String it = get_or_attr(context, section_name, null);
+            String it = get_or_attr(context, section_name);
             boolean field_is_empty =  it == null || TextUtils.isEmpty(Utils.stripHTMLMedia(it).trim());
             boolean conditional_is_negative = section.charAt(2) == '^';
             // Showing inner content if either field is empty and the
@@ -235,16 +229,14 @@ public class Template {
         // For type:, we return directly since no other mod than cloze (or other
         // pre-defined mods) can be present and those are treated separately
         Collections.reverse(mods);
-        Collections.sort(mods, new Comparator<String>() {
-            // This comparator ensures "type:" mods are ordered first in the list. The rest of
-            // the list remains in the same order.
-            @Override
-            public int compare(String lhs, String rhs) {
-                if ("type".equals(lhs)) {
-                    return 0;
-                } else {
-                    return 1;
-                }
+        // This comparator ensures "type:" mods are ordered first in the list. The rest of
+        // the list remains in the same order.
+        //noinspection ComparatorMethodParameterNotUsed
+        Collections.sort(mods, (lhs, rhs) -> {
+            if ("type".equals(lhs)) {
+                return 0;
+            } else {
+                return 1;
             }
         });
 
@@ -378,13 +370,12 @@ public class Template {
         boolean in_mathjax = false;
 
         // The following regex matches one of 3 things, noted below:
-        String regex = new StringBuilder("(?si)")
-            .append("(\\\\[(\\[])|")  // group 1, MathJax opening
-            .append("(\\\\[\\])])|")  // group 2, MathJax close
-            .append("(")              // group 3, Cloze deletion number `ord`
-            .append(String.format(Locale.US, creg, ord))
-            .append(")")
-            .toString();
+        String regex = "(?si)" +
+                "(\\\\[(\\[])|" +  // group 1, MathJax opening
+                "(\\\\[])])|" +  // group 2, MathJax close
+                "(" +              // group 3, Cloze deletion number `ord`
+                String.format(Locale.US, creg, ord) +
+                ")";
 
         Matcher m = Pattern.compile(regex).matcher(txt);
 
@@ -392,12 +383,12 @@ public class Template {
         while (m.find()) {
             if (m.group(1) != null) {
                 if (in_mathjax) {
-                    Log.d(TAG, "MathJax opening found while already in MathJax");
+                    Timber.d("MathJax opening found while already in MathJax");
                 }
                 in_mathjax = true;
             } else if (m.group(2) != null) {
                 if (!in_mathjax) {
-                    Log.d(TAG, "MathJax close found while not in MathJax");
+                    Timber.d("MathJax close found while not in MathJax");
                 }
                 in_mathjax = false;
             } else if (m.group(3) != null) {
@@ -405,13 +396,13 @@ public class Template {
                     // appendReplacement has an issue with backslashes, so...
                     m.appendReplacement(
                         repl,
-                        m.quoteReplacement(
+                        Matcher.quoteReplacement(
                             m.group(0).replace(
                                 "{{c" + ord + "::", "{{C" + ord + "::")));
                     continue;
                 }
             } else {
-                Log.d(TAG, "Unexpected: no expected capture group is present");
+                Timber.d("Unexpected: no expected capture group is present");
             }
             // appendReplacement has an issue with backslashes, so...
             m.appendReplacement(repl, Matcher.quoteReplacement(m.group(0)));

--- a/AnkiDroid/src/main/java/com/ichi2/libanki/template/Template.java
+++ b/AnkiDroid/src/main/java/com/ichi2/libanki/template/Template.java
@@ -71,7 +71,6 @@ public class Template {
     private String mTemplate;
     private Map<String, String> mContext;
 
-    private static String ALT_HANDLEBAR_DIRECTIVE = "{{=<% %>=}}";
 
     private static String get_or_attr(Map<String, String> obj, String name) {
         if (obj.containsKey(name)) {
@@ -155,17 +154,34 @@ public class Template {
         return sb.toString();
     }
 
+    /**
+     * Expands all tags, iteratively until all tags (even tags that are replaced by tags) are resolved.
+     */
+    private String render_tags(@NonNull String template, @NonNull Map<String, String> context) {
+        /* Apply render_some_tags to the tags, until
+           render_some_tags states that it does not find tags to replace anymore
+           anymore. Return the last template state */
+        String previous_template = null;
+        while (template != null) {
+            previous_template = template;
+            template = render_some_tags(template, context);
+        }
+        return previous_template;
+    }
 
     /**
-     * Renders all the tags in a template for a context.
+     * Replaces all the tags in a template in a single pass for the values in the given context map.
      */
-    private String render_tags(String template, Map<String, String> context) {
-        if (template.indexOf(ALT_HANDLEBAR_DIRECTIVE) != -1) {
+    private String render_some_tags(String template, Map<String, String> context) {
+        String ALT_HANDLEBAR_DIRECTIVE = "{{=<% %>=}}";
+        if (template.contains(ALT_HANDLEBAR_DIRECTIVE)) {
             template = template.replace(ALT_HANDLEBAR_DIRECTIVE, "").replace("<%", "{{").replace("%>", "}}");
         }
         StringBuffer sb = new StringBuffer();
         Matcher match = sTag_re.matcher(template);
+        boolean found = false;
         while (match.find()) {
+            found = true;
             String tag_type = match.group(1);
             String tag_name = match.group(2).trim();
             String replacement;
@@ -179,6 +195,9 @@ public class Template {
                 return "{{invalid template}}";
             }
             match.appendReplacement(sb, Matcher.quoteReplacement(replacement));
+        }
+        if (!found) {
+            return null;
         }
         match.appendTail(sb);
         return sb.toString();

--- a/AnkiDroid/src/test/java/com/ichi2/libanki/template/TemplateTest.java
+++ b/AnkiDroid/src/test/java/com/ichi2/libanki/template/TemplateTest.java
@@ -34,6 +34,17 @@ import static org.hamcrest.Matchers.not;
 public class TemplateTest extends RobolectricTest {
 
     @Test
+    public void typeInFieldRenders() {
+        HashMap<String, String> context = new HashMap<>();
+        context.put("Front", "AA{{type:Back}}");
+        Template t = new Template("{{Front}}", context);
+
+        String rendered = t.render();
+
+        assertThat(rendered, is("AA[[type:Back]]"));
+    }
+
+    @Test
     public void testNotFoundWillRender() {
         String maybeBad = "{{#NotFound}}{{NotFound}}{{/NotFound}}";
 

